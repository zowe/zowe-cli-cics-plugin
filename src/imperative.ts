/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

// Imperative version of Zowe CLI
import { IImperativeConfig } from "@brightside/imperative";
import { PluginConstants } from "./api/constants/PluginConstants";
import { CicsSession } from "./cli/CicsSession";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    rootCommandDescription: PluginConstants.PLUGIN_DESCRIPTION,
    productDisplayName: PluginConstants.PLUGIN_NAME,
    name: PluginConstants.PLUGIN_GROUP_NAME,
    pluginHealthCheck: "./lib/healthCheck.handler",
    profiles: [
        {
            type: "cics",
            schema: {
                type: "object",
                title: "CICS Profile",
                description: "A cics profile is required to issue commands in the cics command group that interact with " +
                    "CICS regions. The cics profile contains your host, port, user name, and password " +
                    "for the IBM CICS management client interface (CMCI) server of your choice.",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["H"],
                            description: "The CMCI server host name",
                            type: "string",
                            required: true,
                        },
                    },
                    port: {
                        type: "number",
                        optionDefinition: {
                            name: "port",
                            aliases: ["P"],
                            description: "The CMCI server port",
                            type: "number",
                            defaultValue: 1490,
                        },
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "user",
                            aliases: ["u"],
                            description: "Your username to connect to CICS",
                            type: "string",
                            implies: ["password"],
                            required: true,
                        },
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "password",
                            aliases: ["p"],
                            description: "Your password to connect to CICS",
                            type: "string",
                            implies: ["user"],
                            required: true,
                        },
                    },
                    regionName: {
                        type: "string",
                        optionDefinition: {
                            name: "region-name",
                            description: "The name of the CICS region name to interact with",
                            type: "string"
                        },
                    },
                    cicsPlex: {
                        type: "string",
                        optionDefinition: {
                            name: "cics-plex",
                            description: "The name of the CICSPlex to interact with",
                            type: "string"
                        },
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: CicsSession.CICS_OPTION_REJECT_UNAUTHORIZED
                    },
                    protocol: {
                        type: "string",
                        optionDefinition: CicsSession.CICS_OPTION_PROTOCOL
                    }
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "cics123 --host zos123 --port 1490 --user ibmuser --password myp4ss",
                    description: "Create a cics profile named 'cics123' to connect to CICS at host zos123 and port 1490"
                }
            ]
        }
    ]
};

export = config;
