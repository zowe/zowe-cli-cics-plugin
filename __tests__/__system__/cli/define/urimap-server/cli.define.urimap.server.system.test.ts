/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ITestEnvironment, TestEnvironment, runCliScript } from "@zowe/cli-test-utils";
import { ITestPropertiesSchema } from "../../../../__src__/doc/ITestPropertiesSchema";
import { Session } from "@zowe/imperative";

let TEST_ENVIRONMENT: ITestEnvironment<ITestPropertiesSchema>;
let regionName: string;
let csdGroup: string;
let host: string;
let port: number;
let user: string;
let password: string;
let protocol: string;
let rejectUnauthorized: boolean;
let session: Session;

describe("CICS define urimap-server command", () => {

    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "define_urimap_server",
            installPlugin: true,
            tempProfileTypes: ["cics"]
        });
        const cicsProperties = TEST_ENVIRONMENT.systemTestProperties.cics;
        csdGroup = TEST_ENVIRONMENT.systemTestProperties.cmci.csdGroup;
        regionName = TEST_ENVIRONMENT.systemTestProperties.cmci.regionName;
        host = TEST_ENVIRONMENT.systemTestProperties.cics.host;
        port = TEST_ENVIRONMENT.systemTestProperties.cics.port;
        user = TEST_ENVIRONMENT.systemTestProperties.cics.user;
        password = TEST_ENVIRONMENT.systemTestProperties.cics.password;
        protocol = TEST_ENVIRONMENT.systemTestProperties.cics.protocol;
        rejectUnauthorized = TEST_ENVIRONMENT.systemTestProperties.cics.rejectUnauthorized;
        session = new Session({
            type: "basic",
            hostname: cicsProperties.host,
            port: cicsProperties.port,
            user: cicsProperties.user,
            password: cicsProperties.password,
            rejectUnauthorized: cicsProperties.rejectUnauthorized || false,
            protocol: cicsProperties.protocol as any || "https",
        });
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    it("should be able to display the help", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server_help.sh", TEST_ENVIRONMENT, []);
        expect(output.stderr.toString()).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toMatchSnapshot();
    });

    it("should get a syntax error if urimap name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["", "FAKEGRP", "FAKEPATH", "FAKEHOST", "FAKEPGM", "FAKERGN", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("urimap");
        expect(stderr).toContain("name");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if CSD group is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["FAKESRV", "", "FAKEPATH", "FAKEHOST", "FAKEPGM", "FAKERGN", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("csdGroup");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if urimap path is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["FAKESRV", "FAKEGRP", "", "FAKEHOST", "FAKEPGM", "FAKERGN", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("urimap-path");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if urimap host is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["FAKESRV", "FAKEGRP", "FAKEPATH", "", "FAKEPGM", "FAKERGN", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("urimap-host");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if program name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["FAKESRV", "FAKEGRP", "FAKEPATH", "FAKEHOST", "", "FAKERGN", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("program-name");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if region name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_urimap_server.sh", TEST_ENVIRONMENT,
            ["FAKESRV", "FAKEGRP", "FAKEPATH", "FAKEHOST", "FAKEPGM", "", "false", "TESTSVC"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("region-name");
        expect(output.status).toEqual(1);
    });

});
