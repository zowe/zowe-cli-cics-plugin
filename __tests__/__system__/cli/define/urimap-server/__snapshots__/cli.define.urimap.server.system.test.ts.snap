// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CICS define urimap-server command should be able to display the help 1`] = `
"
 COMMAND NAME
 ------------

   urimap-server | us

 DESCRIPTION
 -----------

   Define a new URIMAP of type server to CICS. This acts as an HTTP(S) server

 USAGE
 -----

   zowe cics define urimap-server <urimapName> <csdGroup> [options]

 POSITIONAL ARGUMENTS
 --------------------

   urimapName		 (string)

      The name of the URIMAP to create. The maximum length of the urimap name is eight
      characters.

   csdGroup		 (string)

      The CICS system definition (CSD) Group for the new urimap that you want to
      define. The maximum length of the group name is eight characters.

 REQUIRED OPTIONS
 ----------------

   --urimap-path | --up (string)

      The path component of the URI.

   --urimap-host | --uh (string)

      The host component of the URI.

   --program-name | --pn (string)

      The application program that makes or handles the requests.

 OPTIONS
 -------

   --urimap-scheme | --us (string)

      The scheme component to be used with the request (http or https).

      Default value: https
      Allowed values: http, https

   --description | --desc (string)

      Description of the URIMAP resource being defined.

   --tcpipservice | --tcpip (string)

      The TCPIPSERVICE to which the URIMAP definition applies.

   --region-name (string)

      The CICS region name to which to define the new URIMAP.

   --cics-plex (string)

      The name of the CICSPlex to which to define the new URIMAP.

   --enable (boolean)

      Whether or not the URIMAP is to be enabled on install by default.

      Default value: true

 CICS CONNECTION OPTIONS
 -----------------------

   --host | -H (string)

      The CICS server host name.

   --port | -P (number)

      The CICS server port.

      Default value: 443

   --user | -u (string)

      Mainframe (CICS) user name, which can be the same as your TSO login.

   --password | --pw (string)

      Mainframe (CICS) password, which can be the same as your TSO password.

   --reject-unauthorized | --ru (boolean)

      Reject self-signed certificates.

      Default value: true

   --protocol | -o (string)

      Specifies CMCI protocol (http or https).

      Default value: https
      Allowed values: http, https

 PROFILE OPTIONS
 ---------------

   --cics-profile | --cics-p (string)

      The name of a (cics) profile to load for this command execution.

   --base-profile | --base-p (string)

      The name of a (base) profile to load for this command execution.

 BASE CONNECTION OPTIONS
 -----------------------

   --token-type | --tt (string)

      The type of token to get and use for the API. Omit this option to use the
      default token type, which is provided by 'zowe auth login'.

   --token-value | --tv (string)

      The value of the token to pass to the API.

   --cert-file (local file path)

      The file path to a certificate file to use for authentication

   --cert-key-file (local file path)

      The file path to a certificate key file to use for authentication

 GLOBAL OPTIONS
 --------------

   --response-format-json | --rfj (boolean)

      Produce JSON formatted data from a command

   --help | -h (boolean)

      Display help text

   --help-examples (boolean)

      Display examples for all the commands in a group

   --help-web | --hw (boolean)

      Display HTML help in browser

 EXAMPLES
 --------

   - Define a URIMAP named URIMAPA for the program named PGM123
   to the region named MYREGION in the CSD group MYGRP where the host is
   www.example.com and the path is /example/index.html:

      $ zowe cics define urimap-server URIMAPA MYGRP --urimap-path /example/index.html --urimap-host www.example.com --program-name PGM123 --region-name MYREGION

"
`;
